CNF-13731: Cert Manager HTTP01 Proxy (#1773)
openshift/enhancements: enhancements/cert-manager-proxy/http01_challenge_proxy.md

---
title: http01-challenge-cert-manager-proxy
authors:
  - "@sebrandon1"
reviewers:
  - "@TrilokGeer"
  - "@swagosh"
approvers:
  - "@tkashem"
  - "@deads2k"
  - "@derekwaynecarr"
api-approvers:
  - "@JoelSpeed"
creation-date: 2025-03-28
last-updated: 2025-03-28
status: implementable
tracking-link:
  - https://issues.redhat.com/browse/CNF-13731
---

# HTTP01 Challenge Proxy for Cert Manager

![HTTP01 Challenge Proxy Diagram](http01_challenge.png)

## Summary

> For baremetal platforms only.  Provide a way for cert-manager to manage certifications for API endpoints (such as api.cluster.example.com) similar to the way it handles certificates for other OpenShift Ingress endpoints.

Reviewer: ```suggestion
For baremetal platforms only.  Provide a way for cert-manager to complete http01 challenges against API endpoints (such as api.cluster.example.com) similar to the way it handles certificate challenges for other OpenShift Ingress endpoints.
```

## Motivation

Cert manager can be used to issue certificates for the OpenShift Container Platform (OCP) endpoints (e.g., console, downloads, oauth) using an external ACME Certificate Authority (CA). These endpoints are exposed via the OpenShift Ingress (`*.apps.cluster.example.com`), and this is a supported and functional configuration today.

> However, cluster administrators often want to issue custom certificates for the API endpoint (`api.cluster.example.com`). Unlike other endpoints, this API endpoint is not exposed via the OpenShift Ingress. Depending on the OCP topology (e.g., SNO, MNO, Compact), it is exposed directly on the node or via a keepalive VIP. This lack of management by the OpenShift Ingress introduces challenges in obtaining certificates using an external ACME CA.

Reviewer: ```suggestion
However, cluster administrators often want to use Cert Managrer to issue custom certificates for the API endpoint (`api.cluster.example.com`). Unlike other endpoints, this API endpoint is not exposed via the OpenShift Ingress. Depending on the OCP topology (e.g., SNO, MNO, Compact), it is exposed directly on the node or via a keepalive VIP. This lack of management by the OpenShift Ingress introduces challenges in obtaining certificates using an external ACME CA.
```

The gap arises due to how the ACME HTTP01 challenge works. The following scenarios illustrate the challenges:

> 1. **SNO (Single Node OpenShift)**: The same nodes host both the ingress and API components. Both FQDNs (`api` and wildcard) resolve to the same IP, making the challenge feasible.
> 2. **Compact Clusters**: The node hosting the API VIP may also host an OpenShift Router. If no router is present on the node hosting the VIP, the challenge will fail.

Reviewer: Missing Standard Cluster topology which is the primary case which fails. Also switched the order to lead with the main issue and then clarify other topologies.
```suggestion
1. **Standard Clusters**: The API VIP is hosted on the control plane nodes which do not host an OpenShift Router. The http01 challenge, which is directed at the API VIP (the IP where `api.cluster.example.com` DNS resolves), will not hit an OpenShift Router and thus not reach the challenge response pod started by Cert Manager.
2. **Compact Clusters**: The node hosting the API VIP may also host an OpenShift Router. If no router is present on the node hosting the VIP, the challenge will fail.
3. **SNO (Single Node OpenShift)**: The same nodes host both the ingress and API components. Both FQDNs (`api` and wildcard) resolve to the same IP, making the challenge feasible.
```

> To address this gap, a small proxy was developed. This proxy runs on the cluster as a DaemonSet and ensures that connections reaching the API on port 80 are redirected to the OpenShift Ingress Routers. The proxy implementation uses `nftables` to redirect traffic from `API:80` to `PROXY:8888`.

Reviewer: ```suggestion
To address this gap, a small proxy was developed. This proxy runs on the cluster as a DaemonSet (control plane nodes) and ensures that connections reaching the API on port 80 are redirected to the OpenShift Ingress Routers. The proxy implementation creates a reverse proxy to the apps VIP and uses `nftables` to redirect traffic from `API:80` to `PROXY:8888`.
```

- **Proxy Code**: [GitHub Repository](https://github.com/mvazquezc/cert-mgr-http01-proxy/tree/main)
- **Deployment Manifest**: [Manifest Link](https://github.com/mvazquezc/cert-mgr-http01-proxy/blob/main/manifests/deploy-in-ocp.yaml)

This enhancement aims to provide a robust solution for managing certificates for the API endpoint in baremetal environments.

### User Stories

> 1. **As a cluster administrator**, I want to issue custom certificates for the API endpoint (`api.cluster.example.com`) using an external ACME CA, so that I can ensure secure communication for my cluster's API.

Reviewer: ```suggestion
1. **As a cluster administrator**, I want to manage custom certificates for the API endpoint (`api.cluster.example.com`) using an external ACME CA, so that I can ensure secure communication for my cluster's API.
```

2. **As a cluster administrator on a baremetal platform**, I want a reliable solution to handle HTTP01 challenges for the API endpoint, even when the endpoint is not managed by OpenShift Ingress, so that I can avoid manual workarounds.
3. **As a developer**, I want a simple deployment mechanism for the HTTP01 challenge proxy, so that I can easily integrate it into my existing cluster setup.

### Proposal

The HTTP01 Challenge Proxy will be implemented via DaemonSet running on the cluster. It will:

- Redirect HTTP traffic from the API endpoint (`api.cluster.example.com`) on port 80 to the OpenShift Ingress Routers.
- Use `nftables` for traffic redirection from `API:80` to `PROXY:8888`.
- Be deployed using a manifest that includes all necessary configurations.

The proxy will ensure compatibility with various OCP topologies, including SNO, MNO, and Compact clusters, addressing the challenges of HTTP01 validation for the API endpoint.

### API Extensions

> A new CR type may be created and can be applied to clusters.  This new typed will be stored in the [openshift/api](https://github.com/openshift/api) repo.

Reviewer: Is a new CR required?

Author: This is something that I discussed as a possibility with @mvazquezc that it would be possibly a new CR that you would apply to get the proxy applied to the nodes.

Reviewer 2: What group will this new CRD be created under? All new APIs will need to be tied to a TPNU feature-gate and start as `v1alpha1`.

What is this API going to look like?

Author: PR opened here: https://github.com/openshift/api/pull/2318

And I will add an example CR in this section.

Author: CR has been added.  I modified the `spec` section per a comment from @imiller0 a while ago, which basically makes it an empty section.

Author: @imiller0 @mvazquezc Correct me if I'm wrong but I believe this should be cluster wide correct?  I can adjust the CR in this case.

Reviewer 3: +1 to cluster wide.

### Implementation Details/Notes/Constraints

> - The proxy will be deployed as a DaemonSet to ensure it runs on all nodes in the cluster.

Reviewer: ```suggestion
- The proxy will be deployed as a DaemonSet to ensure it runs on all nodes which may host the API VIP in the cluster.
```

- The nftables rules will be added to the nodes. The proxy will listen on port 8888 and redirect traffic to the OpenShift Ingress Routers.
- The implementation relies on `nftables` for traffic redirection, which must be supported and enabled on the cluster nodes.
- The demo deployment manifest for the proxy is available [here](https://github.com/mvazquezc/cert-mgr-http01-proxy/blob/main/manifests/deploy-in-ocp.yaml).
- An example implementation can be found in this [repository](https://github.com/mvazquezc/cert-mgr-http01-proxy/tree/main).

### Design Details

- **Proxy Deployment**: The proxy will be deployed using a Kubernetes DaemonSet. The daemonset will implement an nftable rule via pod that runs to completion.
- **Traffic Redirection**: This will use `nftables` rules to redirect incoming traffic on `API:80` to `PROXY:8888`.
- **Security**: The proxy will only handle HTTP traffic for the HTTP01 challenge and will not interfere with other traffic or services.
- **Monitoring**: Logs and metrics will be exposed to help administrators monitor the proxy's behavior and troubleshoot issues.

### Drawbacks

> 1. **Dependency on nftables**: The solution relies on `nftables`, which may not be available or enabled on all environments.

Reviewer: nftables is available on control plane (and worker) nodes.

Reviewer 2: It is _always_ available? What is the expected behavior if it is not available but someone attempts to use this feature?

Author: Might be a better question for @imiller0 or @mvazquezc if the nftables doesn't exist.  I can edit the enhancement proposal with what is discussed.

Reviewer 3: If I'm not mistaken, the nftables are always available. May be worth double-checking with SDN team but iirc the SDN relies on nftables as well.

- An example implementation can be found in this [repository](https://github.com/mvazquezc/cert-mgr-http01-proxy/tree/main).

### Design Details

- **Proxy Deployment**: The proxy will be deployed using a Kubernetes DaemonSet. The daemonset will implement an nftable rule via pod that runs to completion.
- **Traffic Redirection**: This will use `nftables` rules to redirect incoming traffic on `API:80` to `PROXY:8888`.
- **Security**: The proxy will only handle HTTP traffic for the HTTP01 challenge and will not interfere with other traffic or services.
- **Monitoring**: Logs and metrics will be exposed to help administrators monitor the proxy's behavior and troubleshoot issues.

### Drawbacks

1. **Dependency on nftables**: The solution relies on `nftables`, which may not be available or enabled on all environments.
2. **Additional Resource Usage**: Running the proxy as a DaemonSet introduces additional resource usage on the cluster nodes while the proxy pod is applying its nftable rules.
3. **Complexity**: The solution adds another component to the cluster, which may increase operational complexity.

### Alternatives

> None

Reviewer 2: There are _no_ alternatives to this approach that you've considered and ruled out?

Author: I didn't list them here but @mvazquezc put together some alternative proposals in this document (linked in the Jira): https://docs.google.com/presentation/d/1mJ1pnsPiEwb-U5lHwhM2UkyRmkkLeYxj3cfE4F7dOx0/edit#slide=id.g547716335e_0_260

Should I list these other proposals that aren't the PoC's implementation here?

Reviewer 2: My understanding is that EPs are generally used as the source of truth when making changes like this across OCP and are linked to when creating feature gates.

I would at least include a high-level summary of each of the alternative approaches and why they were insufficient.

- Ensure compatibility with various OpenShift topologies, including Standard Clusters, Compact Clusters, and SNO.
- Minimize operational complexity by using a DaemonSet-based deployment and `nftables` for traffic redirection.

## Proposal

The HTTP01 Challenge Proxy will be implemented via DaemonSet running on the cluster. It will:

- Redirect HTTP traffic from the API endpoint (`api.cluster.example.com`) on port 80 to the OpenShift Ingress Routers.
- Use `nftables` for traffic redirection from `API:80` to `PROXY:8888`.
- Be deployed using a manifest that includes all necessary configurations.

The proxy will ensure compatibility with various OCP topologies, including SNO, MNO, and Compact clusters, addressing the challenges of HTTP01 validation for the API endpoint.

### API Extensions

A new CR type may be created and can be applied to clusters.  This new typed will be stored in the [openshift/api](https://github.com/openshift/api) repo.

Potential Example of a CR:

```
apiVersion: network.openshift.io/v1alpha1
kind: HTTP01ChallengeProxy
metadata:
  name: example-http01challengeproxy
  namespace: default
spec:

>   # Add fields here to specify the desired state of the HTTP01ChallengeProxy

Reviewer 2: The motivation and non-goals states only one endpoint, `api.cluster.example.com`, is scoped for this work. Forward-thinking, is there any other endpoints that users may want to be able to proxy that we would want to support?

What other things might a user want to configure here, either now or long-term?

Reviewer 3: That's the only user-exposed endpoint exposed by the platform that isn't routed through the ingress controller, I don't think we need to support other endpoints.

Reviewer 2: One thing I noticed is that there is specification around the proxy being hardcoded to port `8888` - why this port specifically? How should users handle a scenario where something else that is critical to their system needs to run on port `8888`? Would it be worth making the port a configurable value?

status:
  conditions:
    - type: Ready
      status: "True"
      lastTransitionTime: "2025-05-12T00:00:00Z"
      reason: "Initialized"
      message: "HTTP01ChallengeProxy is ready"
```

> ### Implementation Details/Notes/Constraints

Reviewer 3: Answering only the last part, will let other chime in for the first part.

No, multiple instances of a proxy are not allowed to run. There will be port collisions since the port used (8888) will be listening on the host network namespace. This can be mitigated by checking if the port is in use when initializing the proxy. During upgrades, etc, the idea is use the `recreate` strategy so the existing proxy gets deleted before the new one gets created.

> ### Implementation Details/Notes/Constraints

Reviewer 2: How do you plan to limit that only a singular instance of a proxy should run? Is this one proxy per node? one proxy per cluster?

If there is risk of port collision, would it make sense to allow configuration of the port from the user perspective?

> ### Implementation Details/Notes/Constraints

Reviewer 2: It certainly doesn't have to be a full ClusterOperator or a standalone OLM operator. You just need to further specify in this enhancement what component(s) will be responsible for ensuring the proxy gets stamped out onto the cluster and under which conditions.

> - The proxy will be deployed as a DaemonSet to ensure it runs on all nodes which may host the API VIP in the cluster.

Author: @mvazquezc can correct me if I'm wrong but this will be an optional feature that Telco customers can enable.

> - The proxy will be deployed as a DaemonSet to ensure it runs on all nodes which may host the API VIP in the cluster.

Reviewer 2: What approaches have you considered for allowing this to be an opt-in feature?

> - The proxy will be deployed as a DaemonSet to ensure it runs on all nodes which may host the API VIP in the cluster.

Author: I'm assuming we need to agree on how/who/what deploys the daemonset here before the enhancement is finalized right?

> - The proxy will be deployed as a DaemonSet to ensure it runs on all nodes which may host the API VIP in the cluster.

Reviewer 2: Yes.

Depending on how you plan to deploy it may dictate what changes to existing APIs you may make.

For example, if you were to make it the responsibility of an existing cluster operator to deploy this proxy you'd probably want to modify the configuration API it uses as opposed to introducing an entirely new API for it to be responsible for as well.

> - The proxy will be deployed as a DaemonSet to ensure it runs on all nodes which may host the API VIP in the cluster.

Reviewer 3: @JoelSpeed Maybe a better wording would've been "baremetal customers using non cloud-ready DNS servers". For clusters running on cloud providers, the usual configuration is leveraging cloud provider DNS services which can be directly integrated with CertManager to issue certs using the DNS01 challenge.

@everettraven since this is kind of a sidecar for the APIServer, we could use the apiserver CRD. But I'll let @sebrandon1 decide if going with a new cluster operator or if re-using the apiserver CRD.

> - The proxy will be deployed as a DaemonSet to ensure it runs on all nodes which may host the API VIP in the cluster.

Author: Addressing the language in an updated commit.

> - The nftables rules will be added to the nodes. The proxy will listen on port 8888 and redirect traffic to the OpenShift Ingress Routers.

Reviewer 2: >That's done via code.

Please explain, in sentences, the steps that will be done via code to do this. What interactions with other components will be needed? What considerations are in place for different scenarios?

The enhancement should be the source of record for decisions being made. Code changes, even PoC code and that may make future readers of the enhancement not able to follow along as easily with the historical decisions made. Readers of enhancements shouldn't have to dig through code anyways to get an understanding of what is happening with the system as part of this proposal.

> - The nftables rules will be added to the nodes. The proxy will listen on port 8888 and redirect traffic to the OpenShift Ingress Routers.

Reviewer 3: When the proxy starts, it checks which version of OCP it's running on. If it's 4.17+ it will proceed to configure MCO restart the nftables service rather than reboot the node when the file `/etc/sysconfig/nftables.conf` is modified. After that it creates the MC that configures the file + service.

> - The implementation relies on `nftables` for traffic redirection, which must be supported and enabled on the cluster nodes.

Reviewer 3: AFAIK nftables is required for OCP nodes to run. OVN-K requires it (maybe other CNI don't), but still I don't think it can be disabled.

> - The implementation relies on `nftables` for traffic redirection, which must be supported and enabled on the cluster nodes.

Reviewer 2: Maybe @yuqi-zhang ?

> - The implementation relies on `nftables` for traffic redirection, which must be supported and enabled on the cluster nodes.

Reviewer 4: The MCO doesn't directly manage `nftables` in any way on the nodes. I believe it's being shipped (and mostly configured) in RHCOS directly.

As a user, you can technically do whatever you want, including disabling nftables via a MachineConfig, but that would be an explicit user-driven decision and not something we manage.

> - The implementation relies on `nftables` for traffic redirection, which must be supported and enabled on the cluster nodes.

Reviewer 2: Thanks for the additional information @yuqi-zhang !

With that in mind - and no other information provided on it being required - I'd like to have included in the considerations of this EP what should happen if `nftables` isn't present or is disabled and how we plan to detect that.

> - The implementation relies on `nftables` for traffic redirection, which must be supported and enabled on the cluster nodes.

Reviewer 5: Nftables is always present, it is part of the RHCOS payload.

Baremetal and AWS (at least) OCP clusters install nftables tables, chains and rules. Execute `$ nft list ruleset` on any node to list rules. OVN-Kubernetes relies on netfilter (ipables/nftables) for features like UDN, Egress and Services.

In this thread, it is mentioned a few times nftables being enabled/disabled. Note that the `nftables` systemd unit is disabled out of the box *but* netfilter is on and can be configured via the `nftables` CLI/API without the `nftables` systemd unit being enabled.

The `nftables` systemd service is just a helper to faciliate the load/unload of custom firewall rules:

```
[Unit]
Description=Netfilter Tables
Documentation=man:nft(8)
Wants=network-pre.target
Before=network-pre.target

[Service]
Type=oneshot
ProtectSystem=full
ProtectHome=true
ExecStart=/sbin/nft -f /etc/sysconfig/nftables.conf
ExecReload=/sbin/nft 'flush ruleset; include "/etc/sysconfig/nftables.conf";'
ExecStop=/sbin/nft flush ruleset
RemainAfterExit=yes

[Install]
WantedBy=multi-user.target
```

3. **Cert Manager Controller per Spoke**: A Cert Manager controller is configured for each spoke cluster on the hub. These controllers run in the spoke cluster namespace and are configured to use the spoke’s `system:admin` kubeconfig. This approach allows centralized control but requires managing multiple controllers on the hub.

4. **Single Cert Manager Controller on the Hub**: A single Cert Manager controller runs on the hub. Certificates and `CertificateRequests` for each spoke cluster are created with data known beforehand (e.g., API, Ingress, CNFs). The resulting secrets are synced to the spokes via RHACM Policies. This approach simplifies the deployment but requires pre-configured data for each spoke.

More information about the investigation can be found [here](https://docs.google.com/presentation/d/1mJ1pnsPiEwb-U5lHwhM2UkyRmkkLeYxj3cfE4F7dOx0/edit#slide=id.g547716335e_0_260).

### Risks and Mitigations

> 1. **Proxy Failure**: If the proxy fails, HTTP01 challenges for the API endpoint will not succeed. Mitigation: Use health checks and monitoring to ensure the proxy is running correctly.

Reviewer 2: If possible, please enumerate the impact, if any, to end-users and other system components during this scenario. Include the steps they may be able to take to workaround or remedy the issue.

> 1. **Proxy Failure**: If the proxy fails, HTTP01 challenges for the API endpoint will not succeed. Mitigation: Use health checks and monitoring to ensure the proxy is running correctly.

Author: Adding some wording around this.

> 2. **Traffic Interference**: The proxy could inadvertently interfere with other traffic. Mitigation: Carefully scope the proxy's functionality to only handle HTTP01 challenge traffic.

Reviewer 3: We should configure alerts for expiring certificates. This way we will be alerted if we are close to renewal time.

> 2. **Traffic Interference**: The proxy could inadvertently interfere with other traffic. Mitigation: Carefully scope the proxy's functionality to only handle HTTP01 challenge traffic.

Author: I'll add some verbiage about the above.

## Test Plan

1. **Unit Tests**: Validate the proxy's functionality in isolation, including traffic redirection and error handling.
2. **Integration Tests**: Deploy the proxy in a test cluster and verify that HTTP01 challenges for the API endpoint succeed.
3. **Performance Tests**: Measure the proxy's impact on cluster performance and resource usage.
4. **Topology Tests**: Test the proxy in Standard Clusters, Compact Clusters, and SNO environments to ensure compatibility.

## Graduation Criteria

### Dev Preview -> Tech Preview

- The proxy is implemented and tested in development environments.
- Documentation is available for deploying and configuring the proxy.

### Tech Preview -> GA

- The proxy is deployed in production environments and successfully handles HTTP01 challenges for various OCP topologies.
- Performance and reliability meet production-grade requirements.

### Removing a deprecated feature

This enhancement does not deprecate any existing features.

## Upgrade / Downgrade Strategy

> Updated versions of the proxy can be applied to the cluster similar to initial deployment

Reviewer 2: Beyond the proxy itself, how do you plan to handle typical upgrade situations where there may older and newer instances of the proxy running at the same time due to a rolling upgrade? Are there any special considerations needed for this scenario?

What happens if an upgrade fails midway through? What steps would a user need to take to get their cluster back to the previous state?

Author: Adding additional context here.

## Version Skew Strategy

> Any changes to the proxy's behavior will be documented to ensure compatibility with older cluster versions.

Reviewer 2: Does there need to be any special considerations for the interactions between the component that deploys the proxy and updates node configurations and the component that rolls out node configuration changes?

During upgrades there will always be version skew between components, so capturing how your component handles this is helpful.

Reviewer 2: I think some of the nuance for how you handle version skew is going to largely depend on how the proxy is deployed.

For example, adding logic to an existing cluster operator to manage this means it will only be possible to deploy this proxy starting from the OCP version that the logic is introduced to the cluster operator.

On the flip side, if you are suggesting customers install this via some manifests, you have no control over which version of OCP it would be deployed on and having the proxy identify the OCP version it is running on may be useful.

It's important to first identify how the proxy actually gets on a cluster before thinking through some of these version skew scenarios.

Reviewer 3: @everettraven Currently the proxy code checks which version of OCP it's running on. If the version is <4.17 it gracefully fails to start with an error describing that it's supported in releases 4.17+.

Based on previous comments, the idea is that the proxy gets on a cluster via a cluster operator. The workload object used will be a DaemonSet.

## Operational Aspects of API Extensions

- **Monitoring**: Logs and metrics will be exposed to help administrators monitor the proxy's behavior and troubleshoot issues.
- **Resource Usage**: The proxy's resource requirements will be minimal, as it only handles HTTP01 challenge traffic.

> - **Failure Recovery**: Health checks will ensure that the proxy is running correctly, and failed pods will be automatically restarted.

Author: Okay adding a `Recovery Procedure` section.

## Support Procedures

> Support for the proxy will be provided through standard OpenShift support channels. Administrators can refer to the deployment documentation and logs for troubleshooting.

Reviewer 2: I'd like to see a bit more information here "standard support channels" isn't all that helpful in representing how someone providing support can evaluate if this component is behaving correctly or not and what steps they can take to mitigate issues that it may be causing on the system.

The template section has some good examples for things to take into consideration: https://github.com/openshift/enhancements/blob/master/guidelines/enhancement_template.md#support-procedures

---
title: http01-challenge-cert-manager-proxy
authors:
  - "@sebrandon1"
reviewers:
  - "@TrilokGeer"
  - "@swagosh"
approvers:
  - "@tkashem"
  - "@deads2k"
  - "@derekwaynecarr"
api-approvers:
  - "@JoelSpeed"
creation-date: 2025-03-28
last-updated: 2025-03-28
status: implementable
tracking-link:
  - https://issues.redhat.com/browse/CNF-13731
---

# HTTP01 Challenge Proxy for Cert Manager

![HTTP01 Challenge Proxy Diagram](http01_challenge.png)

## Summary

For baremetal platforms only.  Provide a way for cert-manager to complete http01 challenges against API endpoints (such as api.cluster.example.com) similar to the way it handles certificate challenges for other OpenShift Ingress endpoints.

## Motivation

Cert manager can be used to issue certificates for the OpenShift Container Platform (OCP) endpoints (e.g., console, downloads, oauth) using an external ACME Certificate Authority (CA). These endpoints are exposed via the OpenShift Ingress (`*.apps.cluster.example.com`), and this is a supported and functional configuration today.

> However, cluster administrators often want to use Cert Manager to issue custom certificates for the API endpoint (`api.cluster.example.com`). Unlike other endpoints, this API endpoint is not exposed via the OpenShift Ingress. Depending on the OCP topology (e.g., SNO, MNO, Compact), it is exposed directly on the node or via a keepalive VIP. This lack of management by the OpenShift Ingress introduces challenges in obtaining certificates using an external ACME CA.

Author: @mvazquezc would you like to comment on this?

- **Deployment Manifest**: [Manifest Link](https://github.com/mvazquezc/cert-mgr-http01-proxy/blob/main/manifests/deploy-in-ocp.yaml)

This enhancement aims to provide a robust solution for managing certificates for the API endpoint in baremetal environments.

### User Stories

1. **As a cluster administrator**, I want to manage custom certificates for the API endpoint (`api.cluster.example.com`) using an external ACME CA, so that I can ensure secure communication for my cluster's API.
2. **As a cluster administrator on a baremetal platform**, I want a reliable solution to handle HTTP01 challenges for the API endpoint, even when the endpoint is not managed by OpenShift Ingress, so that I can avoid manual workarounds.
3. **As a developer**, I want a simple deployment mechanism for the HTTP01 challenge proxy, so that I can easily integrate it into my existing cluster setup.

### Goals

- Provide a reliable mechanism for Cert Manager to complete HTTP01 challenges for the API endpoint (`api.cluster.example.com`) in baremetal environments.
- Ensure compatibility with various OpenShift topologies, including Standard Clusters, Compact Clusters, and SNO.

> - Minimize operational complexity by using a DaemonSet-based deployment and `nftables` for traffic redirection.

Author: Verbiage updated.

## Proposal

> The HTTP01 Challenge Proxy will be implemented via DaemonSet running on the cluster. It will:

Reviewer 3: Running on nodes that can host the API VIP.

> - Redirect HTTP traffic from the API endpoint (`api.cluster.example.com`) on port 80 to the OpenShift Ingress Routers.

Reviewer 3: 403 - Forbidden: Only /.well-known/acme-challenge/* is allowed

- Use `nftables` for traffic redirection from `API:80` to `PROXY:8888`.
- Be deployed using a manifest that includes all necessary configurations.

The proxy will ensure compatibility with various OCP topologies, including SNO, MNO, and Compact clusters, addressing the challenges of HTTP01 validation for the API endpoint.

### API Extensions

> A new CR type may be created and can be applied to clusters.  This new typed will be stored in the [openshift/api](https://github.com/openshift/api) repo.

Reviewer 3: In my opinion this makes more sense under the kube api server operator.

- OVN-Kubernetes relies on netfilter (iptables/nftables) for features like UDN, Egress, and Services.
- The Machine Config Operator (MCO) does not manage `nftables` directly, but users can explicitly disable or modify `nftables` via MachineConfig if desired.
- If a user disables or removes `nftables`, this is considered an explicit user-driven action and is not managed or expected by OpenShift.
- If `nftables` is not present or is disabled, the proxy will not function as intended. Detection of this condition should be implemented (e.g., by checking for the presence of the `nft` binary and ability to apply rules), and the operator should surface a clear error or degraded status.
- Based on current RHCOS and OCP design, it is not possible to remove the underlying netfilter subsystem, so the feature can reliably depend on its presence unless a user takes explicit unsupported action.

### Design Details

- **Proxy Deployment**: The proxy will be deployed using a Kubernetes DaemonSet. The daemonset will implement an nftable rule via pod that runs to completion.
- **Traffic Redirection**: This will use `nftables` rules to redirect incoming traffic on `API:80` to `PROXY:8888`.

> - **Security**: The proxy will only handle HTTP traffic for the HTTP01 challenge and will not interfere with other traffic or services.

Reviewer 3: Returns 403 - Forbidden code. If you were to send traffic to the challenge path with non-sense data, it will reach the cert-manager challenge pod (in case there is one) as it would without this proxy.

1. **RHACM Manages Cert Manager Deployment**: RHACM (Red Hat Advanced Cluster Management) manages the deployment of Cert Manager and certificates on the spokes using Policies. Each managed cluster runs its own Cert Manager instance. This approach decentralizes certificate management but requires Cert Manager to be deployed and maintained on each spoke cluster.

2. **Single Addon on the Hub**: A single addon runs on the hub and watches the spoke clusters' APIs for `Certificate` and `CertificateRequest` related events. When these APIs are created, updated, or deleted in the spoke, the addon syncs the contents back and forth between the hub and the spokes. This approach centralizes management but introduces additional complexity in syncing data.

3. **Cert Manager Controller per Spoke**: A Cert Manager controller is configured for each spoke cluster on the hub. These controllers run in the spoke cluster namespace and are configured to use the spoke’s `system:admin` kubeconfig. This approach allows centralized control but requires managing multiple controllers on the hub.

> 4. **Single Cert Manager Controller on the Hub**: A single Cert Manager controller runs on the hub. Certificates and `CertificateRequests` for each spoke cluster are created with data known beforehand (e.g., API, Ingress, CNFs). The resulting secrets are synced to the spokes via RHACM Policies. This approach simplifies the deployment but requires pre-configured data for each spoke.

Reviewer 3: The approaches listed here are related to how to handle cert issuing at scale. All of these approaches will hit the same "gap" that this enhancement aims to address.

> 4. **Single Cert Manager Controller on the Hub**: A single Cert Manager controller runs on the hub. Certificates and `CertificateRequests` for each spoke cluster are created with data known beforehand (e.g., API, Ingress, CNFs). The resulting secrets are synced to the spokes via RHACM Policies. This approach simplifies the deployment but requires pre-configured data for each spoke.

Author: Addressed in the document.

### API Extensions

Rather than creating a new CRD, the HTTP01 Challenge Proxy configuration will be added as a new field to the existing **APIServer** CRD in the [openshift/api](https://github.com/openshift/api) repo. This approach aligns with the pattern used by other API server features like audit, encryption, and TLS configuration.

The configuration will be added to the `APIServerSpec` struct and managed by the cluster-kube-apiserver-operator as part of its normal reconciliation loop. Since the APIServer CRD is already a singleton (named "cluster"), no additional enforcement mechanisms are needed.

**API Changes to config/v1/types_apiserver.go:**

```go
type APIServerSpec struct {
    // ... existing fields ...
    
    // http01ChallengeProxy contains configuration for the HTTP01 challenge proxy
    // that redirects traffic from the API endpoint on port 80 to ingress routers.
    // This enables cert-manager to perform HTTP01 ACME challenges for API endpoint certificates.
    // +optional
    HTTP01ChallengeProxy *HTTP01ChallengeProxySpec `json:"http01ChallengeProxy,omitempty"`
}

> type HTTP01ChallengeProxySpec struct {

Reviewer 2: Because the parent field is already optional, you'll want at least one of these children fields to be required.

    // enabled controls whether the HTTP01 challenge proxy is active.
    // When enabled, the proxy redirects HTTP traffic from API:80 to ingress routers.
    // +optional

>     Enabled bool `json:"enabled"`

Reviewer 2: Do not use booleans. Use an enum instead (ref: https://github.com/openshift/enhancements/blob/master/dev-guide/api-conventions.md#do-not-use-boolean-fields).

You may also want to consider using a discriminated union to allow configuration under certain modes of operation and not allowing configuration under others (ref: https://github.com/openshift/enhancements/blob/master/dev-guide/api-conventions.md#discriminated-unions)

With these in mind, the end-result of your API may look something like:
```yaml
apiVersion: config.openshift.io/v1
kind: APIServer
metadata:
  name: cluster
spec:
  # existing fields
  http01ChallengeProxy:
    mode: { DefaultDeployment || CustomDeployment }
    customDeployment:
      internalPort: 8888
```

This approach also makes it easier to extend this API in the future if the use cases arise because you can add new "deployment modes" and add additional configuration options under them.

    
    // internalPort specifies the internal port used by the proxy service.
    // +kubebuilder:validation:Minimum=1024
    // +kubebuilder:validation:Maximum=65535
    // +optional
    InternalPort int32 `json:"internalPort,omitempty"`  // default: 8888
}
```

**Example Configuration:**

```yaml
apiVersion: config.openshift.io/v1
kind: APIServer
metadata:
  name: cluster
spec:
  # existing fields...
  http01ChallengeProxy:
    enabled: true
    internalPort: 8888
status:
  # existing status fields...
  conditions:
    - type: HTTP01ChallengeProxyReady
      status: "True"
      lastTransitionTime: "2025-05-12T00:00:00Z"
      reason: "Initialized"
      message: "HTTP01ChallengeProxy is ready"
```
    message: "Only one HTTP01ChallengeProxy instance named 'cluster' is allowed per cluster"

> ```

Reviewer 2: Nit: looks like some formatting issues start here due to unclosed code fence?

This design ensures that:
- Only one `HTTP01ChallengeProxy` instance can exist cluster-wide
- The instance must be named `cluster` to enforce the singleton pattern

> - The ValidatingAdmissionPolicy prevents creation of additional instances

Reviewer 2: Where did we specify in the document that a VAP would be included?

Could we get an outline of the VAP we expect to be created?

- The cluster-kube-apiserver-operator manages this singleton instance

### Implementation Details/Notes/Constraints

- The **cluster-kube-apiserver-operator** will be responsible for deploying and managing the proxy as a DaemonSet on control plane nodes that may host the API VIP.
- **Deployment Conditions**: The operator deploys the proxy DaemonSet only when:
  - `APIServer.spec.http01ChallengeProxy.enabled = true` is set in the cluster's APIServer configuration
  - The cluster is running on a baremetal or supported platform (not cloud platforms with integrated DNS01 support)
  - The operator validates that nftables/netfilter subsystem is available on target nodes
- **Component Integration**: The operator will integrate the proxy lifecycle with existing API server management, ensuring consistent behavior and upgrade/downgrade strategies.
- **Singleton Enforcement**: Only one proxy configuration per cluster is supported, enforced by the singleton nature of the APIServer CRD.

> - When the proxy starts, it checks which version of OCP it's running on. If it's 4.17+ it will proceed to configure MCO restart the nftables service rather than reboot the node when the file /etc/sysconfig/nftables.conf is modified. After that it creates the MC that configures the file + service.

Reviewer 2: If this proxy support is being managed by a cluster-operator, it will only ever exist on an OpenShift version that the cluster operator is associated with. (i.e if this feature is added in 4.20, it will only ever exist on 4.20+ clusters.)

I suspect the proxy itself doesn't actually need to be aware of the OCP version it is running on anymore.

Maybe CKASO can be responsible for this re-configuration of the MCO as well?

### Design Details

- **Operator Integration**: The cluster-kube-apiserver-operator will manage the complete lifecycle of the HTTP01 challenge proxy, including deployment, configuration, upgrades, and removal.
- **Proxy Deployment**: The operator will deploy the proxy feature as a DaemonSet on control plane nodes. The daemonset will implement nftable rules via pods that run to completion.
- **Traffic Redirection**: This will use `nftables` rules to redirect incoming traffic on `API:80` to `PROXY:8888`.
- **Security**: The proxy implements selective traffic filtering, only handling HTTP requests with paths matching `/.well-known/acme-challenge/*` destined for the API VIP. 
  - **ACME Traffic**: Requests to `<API_VIP>/.well-known/acme-challenge/*` are redirected to the ingress routers for certificate validation.
    - **Valid ACME Data**: Properly formatted challenge requests are forwarded to cert-manager challenge pods for validation.
    - **Invalid ACME Data**: Malformed challenge requests to the correct path are still forwarded to cert-manager, which handles the validation failure (same behavior as without the proxy).

>   - **Non-ACME Traffic**: HTTP requests to the API VIP on port 80 with other paths receive a **403 Forbidden** response with the message "Only /.well-known/acme-challenge/* is allowed".

Reviewer 2: Would a 400 Bad Request be more appropriate than a 403 Forbidden?

A 403 makes it sound like it is an explicit restriction of permissions for an individual user as opposed to the proxy only understanding how to handle traffic to the `/.well-known/acme-challenge/*` endpoint.

Author: Updated in the doc.

  - **HTTPS Traffic**: All HTTPS traffic (port 443) to the API endpoint continues to function normally and is unaffected by the proxy.
- **Monitoring**: Logs and metrics will be exposed to help administrators monitor the proxy's behavior and troubleshoot issues.

> - **Configuration Management**: The operator will watch the `HTTP01ChallengeProxy` CR and reconcile the proxy state accordingly.

Reviewer 2: This seems like it is currently outdated.

3. **Cert Manager Controller per Spoke**: A Cert Manager controller is configured for each spoke cluster on the hub. These controllers run in the spoke cluster namespace and are configured to use the spoke’s `system:admin` kubeconfig. This approach allows centralized control but requires managing multiple controllers on the hub.

4. **Single Cert Manager Controller on the Hub**: A single Cert Manager controller runs on the hub. Certificates and `CertificateRequests` for each spoke cluster are created with data known beforehand (e.g., API, Ingress, CNFs). The resulting secrets are synced to the spokes via RHACM Policies. This approach simplifies the deployment but requires pre-configured data for each spoke.

More information about the investigation can be found [here](https://docs.google.com/presentation/d/1mJ1pnsPiEwb-U5lHwhM2UkyRmkkLeYxj3cfE4F7dOx0/edit#slide=id.g547716335e_0_260).

> ### Risks and Mitigations

Reviewer 2: Would it be worth noting that just disabling the proxy is also a potential mitigation for these risks?

Author: Adjusted in the doc.

4. **Selective Redirection**: Only traffic matching the ACME challenge path pattern is intercepted using `nftables` and redirected to the proxy pod on port 8888. Other HTTP requests to the API VIP on port 80 receive a **403 Forbidden** response.
5. The proxy pod forwards the filtered request to the OpenShift Ingress Router, which serves the challenge response from the Cert Manager challenge pod.
6. The ACME CA validates the challenge and issues the certificate for the API endpoint.

### Topology Considerations

- **Standard Clusters**: The API VIP is hosted on control plane nodes. The proxy ensures that HTTP01 challenges are redirected to the OpenShift Ingress Routers.
- **Compact Clusters**: The proxy handles scenarios where the API VIP node may or may not host an OpenShift Router, ensuring consistent challenge redirection.
- **SNO (Single Node OpenShift)**: The proxy is not strictly required in this topology, as the API and wildcard FQDNs resolve to the same IP. However, it can still be deployed for consistency.

#### Hypershift / Hosted Control Planes

> This enhancement does not directly apply to Hypershift deployments, as the API endpoint management in Hypershift differs from baremetal environments. However, the proxy's design could be adapted for similar use cases in Hypershift if needed.

Reviewer 2: Because you are making changes to the `apiservers.config.openshift.io` resource this _does_ impact HyperShift as HyperShift's APIs embed the spec for that resource in their resources: https://github.com/openshift/hypershift/blob/d5884761640027a4d4d78fe1711597a9cef867a7/api/hypershift/v1beta1/hostedcluster_types.go#L1822-L1826

If you do not want this feature to apply to HyperShift, how do you intend to prevent the field from being embedded within HyperShift types (or how do you prevent HyperShift users from being able to set it if they can't use it)?

Reviewer 3: We haven't tested in HyperShift, and I don't think the solution will work as it is today on HyperShift since it relies on MCO  that works differently in HyperShift.

Does the KASO know the topology it is running on? in case it does, we can have it ignore this setting in HyperShift topology until we decide how this gets implemented in HyperShift. WDYT?

Author: Pinged @sjenning @csrwng and @enxebre on Slack to comment here.

  - Logs from the proxy pod show errors related to `nftables` or port binding.
  - The API server logs may show failed ACME challenge requests or timeouts.

- **Metrics/Alerts**:
  - Custom metrics (if implemented) such as `cert_manager_proxy_up` or `cert_manager_proxy_errors_total` may indicate proxy health.
  - Alerts can be configured for DaemonSet unavailability or excessive restarts.

### Disabling the API Extension

- **How to disable**: 
  - Remove or scale down the proxy DaemonSet.
  - Remove or update the associated CR (if using a CRD).
  - Remove any MachineConfig or configuration that enables the proxy.
- **Consequences**:
  - HTTP01 challenges for the API endpoint will not be possible.
  - Certificates for the API endpoint will not be issued or renewed.
  - If the API certificate expires, cluster API access may be impacted until a valid certificate is restored.

### Impact on Existing, Running Workloads

- Existing workloads and API traffic will continue to function as long as the API certificate is valid.
- If the API certificate expires and is not renewed, clients may fail to connect to the API server due to certificate errors.

> - No direct impact on running pods or services, unless they rely on the API endpoint with a valid certificate.

Reviewer 2: API endpoint meaning the public route to the API endpoint? Just wanting to verify that this shouldn't have any impact to internal DNS that operator pods, for example, may use.

The configuration will be added to the `APIServerSpec` struct and managed by the cluster-kube-apiserver-operator as part of its normal reconciliation loop. Since the APIServer CRD is already a singleton (named "cluster"), no additional enforcement mechanisms are needed.

**API Changes to config/v1/types_apiserver.go:**

```go
type APIServerSpec struct {
    // ... existing fields ...
    
    // http01ChallengeProxy contains configuration for the HTTP01 challenge proxy
    // that redirects traffic from the API endpoint on port 80 to ingress routers.
    // This enables cert-manager to perform HTTP01 ACME challenges for API endpoint certificates.
    // +optional
    HTTP01ChallengeProxy *HTTP01ChallengeProxySpec `json:"http01ChallengeProxy,omitempty"`
}

type HTTP01ChallengeProxySpec struct {
    // mode controls whether the HTTP01 challenge proxy is active and how it should be deployed.
    // DefaultDeployment enables the proxy with default configuration.
    // CustomDeployment enables the proxy with user-specified configuration.
    // +kubebuilder:validation:Enum="";DefaultDeployment;CustomDeployment

>     // +kubebuilder:validation:Required

Author: Addressed in doc.

    Mode string `json:"mode"`
    
    // customDeployment contains configuration options when mode is CustomDeployment.
    // This field is only valid when mode is CustomDeployment.
    // +optional
    CustomDeployment *HTTP01ChallengeProxyCustomDeploymentSpec `json:"customDeployment,omitempty"`
}

type HTTP01ChallengeProxyCustomDeploymentSpec struct {
    // internalPort specifies the internal port used by the proxy service.
    // +kubebuilder:validation:Minimum=1024

>     // +kubebuilder:validation:Maximum=65535

Author: Addressed in doc.

### Implementation History

- **2025-03-28**: Enhancement proposal created.

### References

- [Cert Manager Expansion JIRA Epic](https://issues.redhat.com/browse/CNF-13731)
- [ACME HTTP01 Challenge](https://letsencrypt.org/docs/challenge-types/#http-01-challenge)
- [Proxy Code Repository](https://github.com/mvazquezc/cert-mgr-http01-proxy/tree/main)
- [Deployment Manifest](https://github.com/mvazquezc/cert-mgr-http01-proxy/blob/main/manifests/deploy-in-ocp.yaml)

### Workflow Description

1. Cert Manager initiates an HTTP01 challenge for the API endpoint (`api.cluster.example.com`).
2. The HTTP01 challenge request is directed to the API VIP on port 80 with the path `/.well-known/acme-challenge/<token>`.
3. **Traffic Identification**: The HTTP01 Challenge Proxy identifies this as ACME challenge traffic by examining the destination path (`/.well-known/acme-challenge/*`) rather than using SNI validation.

> 4. **Selective Redirection**: Only traffic matching the ACME challenge path pattern is intercepted using `nftables` and redirected to the proxy pod on port 8888. Other HTTP requests to the API VIP on port 80 receive a **403 Forbidden** response.

Reviewer 3: We can change this to 400 Bad Request.

## Test Plan

1. **Unit Tests**: Validate the proxy's functionality in isolation, including traffic redirection and error handling.
2. **Integration Tests**: Deploy the proxy in a test cluster and verify that HTTP01 challenges for the API endpoint succeed.
3. **Performance Tests**: Measure the proxy's impact on cluster performance and resource usage.
4. **Topology Tests**: Test the proxy in Standard Clusters, Compact Clusters, and SNO environments to ensure compatibility.

## Graduation Criteria

### Dev Preview -> Tech Preview

- The proxy is implemented and tested in development environments.
- Documentation is available for deploying and configuring the proxy.

### Tech Preview -> GA

- The proxy is deployed in production environments and successfully handles HTTP01 challenges for various OCP topologies.
- Performance and reliability meet production-grade requirements.

### Removing a deprecated feature

This enhancement does not deprecate any existing features.

## Upgrade / Downgrade Strategy

- Updated versions of the proxy can be applied to the cluster similar to initial deployment.

> - The proxy DaemonSet must use a `Recreate` update strategy to ensure that only one instance of the proxy runs per node at any time, as the proxy listens on a fixed port (`8888`) in the host network. This prevents port collisions during upgrades.

Reviewer 6: You're stretching my networking knowledge 😅 I believe no, you wouldn't be able to do that outside of the host network namespace. May just be worth a note in the EP why you need host network (I didn't see that, but it might be there?)

- Low operational complexity to secure communication with a cluster's API endpoint.

### Non-Goals

- This enhancement does not aim to replace or modify the existing OpenShift Ingress functionality.
- It does not provide support for non-HTTP01 challenge types (e.g., DNS-01).
- It does not address certificate management for endpoints other than the API endpoint (`api.cluster.example.com`).
- It does not provide a solution for environments where `nftables` is not supported.

## Proposal

The HTTP01 Challenge Proxy will be implemented as a **core component of the OpenShift payload** and managed by an existing cluster operator, specifically the **cluster-kube-apiserver-operator**, which already manages API endpoint configuration and lifecycle.

**Deployment Architecture:**
- **Core Integration**: This feature is part of the core OpenShift distribution, not an optional OLM-installable operator
- **Operator Management**: The cluster-kube-apiserver-operator will deploy and manage the proxy DaemonSet
- **Conditional Deployment**: The proxy is only deployed when `APIServer.spec.http01ChallengeProxy.mode` is set to `DefaultDeployment` or `CustomDeployment`

> - **Platform Targeting**: Limited to baremetal platforms where HTTP01 challenges are needed for API certificates

Reviewer 7: > https://github.com/openshift/enhancements/pull/1773#discussion_r2215490405
> https://github.com/openshift/enhancements/pull/1773#pullrequestreview-3176944091 

Regarding the use of this feature on non-baremetal platforms, I understand that the motivation and focus are derived from baremetal, but I still believe we shouldn't prevent non-baremetal users from enabling it with additional validations.

As cert-manager users, while major cloud providers already offer DNS provider support, enabling this feature would provide more options. Users could evaluate and adopt it based on their needs, and this would also represent additional flexibility that OpenShift can offer. What do you think? @mvazquezc @JoelSpeed

Reviewer 6: Taking AWS as an example, does the API server load balancer come configured today to forward traffic on port 80 to the control plane nodes? I suspect it does not.

If it does not, enabling this feature on that platform would achieve nothing? The proxy could be exposed and handle traffic incoming on port 80, but there would still be no way an ACME challenge could hit the proxy

So if my assumption there is correct, what would be the point in allowing someone to configure this on AWS?

- If the API certificate has expired, recovery may require connecting to the cluster's API server while ignoring certificate validation errors (e.g., using `--insecure-skip-tls-verify` with `oc` or `kubectl`).
- After resolving the issue (e.g., fixing the DaemonSet, node configuration, or proxy image), re-deploy the proxy to restore HTTP01 challenge functionality.
- The proxy should surface clear status and error messages to help identify and resolve CrashLoopBackOff or degraded states.

## Support Procedures

### Detecting Failure Modes

- **Symptoms**: 
  - Cert Manager HTTP01 challenges for the API endpoint (`api.cluster.example.com`) fail to complete.
  - Certificates for the API endpoint are not issued or renewed.
  - The proxy DaemonSet pods are in `CrashLoopBackOff` or `Error` state.
  - Events in the `openshift-cert-manager` or relevant namespace indicate pod failures.
  - The operator managing the proxy (if any) reports a degraded or error status.
  - Logs from the proxy pod show errors related to `nftables` or port binding.
  - The API server logs may show failed ACME challenge requests or timeouts.

- **Metrics/Alerts**:

>   - Custom metrics (if implemented) such as `cert_manager_proxy_up` or `cert_manager_proxy_errors_total` may indicate proxy health.

Reviewer 7: ```suggestion
  - Custom metrics (if implemented) such as `apiserver_http01_challenge_proxy_up` or `apiserver_http01_challenge_proxy_errors_total` may indicate proxy health.
```

- Redirect filtered HTTP traffic from the API endpoint (`api.cluster.example.com`) on port 80 to the OpenShift Ingress Routers  
- Use `nftables` for traffic redirection from `API:80` to `PROXY:8888`
- Handle lifecycle management, upgrades, and configuration of the proxy

The proxy will ensure compatibility with various OCP topologies, including SNO, MNO, and Compact clusters, addressing the challenges of HTTP01 validation for the API endpoint.

### API Extensions

Rather than creating a new CRD, the HTTP01 Challenge Proxy configuration will be added as a new field to the existing **APIServer** CRD in the [openshift/api](https://github.com/openshift/api) repo. This approach aligns with the pattern used by other API server features like audit, encryption, and TLS configuration.

The configuration will be added to the `APIServerSpec` struct and managed by the cluster-kube-apiserver-operator as part of its normal reconciliation loop. Since the APIServer CRD is already a singleton (named "cluster"), no additional enforcement mechanisms are needed.

**API Changes to config/v1/types_apiserver.go:**

```go

> type APIServerSpec struct {

Author: Section has been updated based on the changes in the PR.
